src/cryptonote_core/blockchain.cpp

```diff
@@ -3552,6 +3553,9 @@ bool Blockchain::handle_block_to_main_chain(const block& bl, const crypto::hash&
   get_difficulty_for_next_block(); // just to cache it
   invalidate_block_template_cache();
 
+  if (m_block_notify)
```

Comment:
```c++
auto notify = m_block_notify.get();
if (notify)
{
  notify->notify(epee::string_tools::pod_to_hex(id).c_str());
}
```

Reply:
Any reason why? shared_ptr has an operator bool which checks if a pointer is set.

Reply:
Thread safety

Reply:
How does `get` have any more thread safety than `operator bool`? Both are thread safe.

Reply:
Please, check the code snippet once again, i updated it as i noted in the previous reply.

Reply:
You're now creating a new shared_ptr (in auto) and then using the bool operator. It's not needed. The original code was just fine as `if (m_block_notify)`.

Reply:
Guess what happens when another thread changes `m_block_notify` in between
```c++
if (m_block_notify)
```
and
```c++
m_block_notify->notify(epee::string_tools::pod_to_hex(id).c_str());
``` 
?

Reply:
I take your point, but if the goal is thread safety to this level, a mutex should really be used on anything relying on `m_block_notify` (or any other member variable for that matter). The overhead of either creating a temporary (as in your example) or using a mutex, is really just unnecessary overhead IMHO.

Reply:
Stability, that's what really matters. Especially in the Monero's case.

Reply:
I look forward to your changing of all member variable access / usage then in this class if you believe this to be unstable. Thanks.

Reply:
Haha, don't be so full of butthurt. Shit happens.

Sure, i will point to any problem i will identify in the code.
That is what is called code review, FYI.

Reply:
Not butthurt at all. Genuinely pleased things get reviewed and debated before they get merged - it's critically important. That's why I submit code, participate in code reviews of others and support the community in all ways I can. 

In the case of this class, it turns out it uses those `CRITICAL_...` macros for locking to get thread safety. For consistency, this variable should ideally follow suit I guess.

Reply:
If we are goin' to keep `shared_ptr` and add a `mutex`, there won't be any point to use `shared_ptr` exactly in this case then. Will have to replace `shared_ptr` with `unique_ptr` and use a `mutex` to maintain `unique_ptr`'s thread-safety.

And that's obviously would be a self-made `shared_ptr` parody. That's what `shared_ptr` actually do.

Reply:
I wonder if the copy operator is atomic...

Reply:
@moneromooo-monero @jtgrassie 
Had to revisit the standard once again. Unfortunately, yes, `std::shared_ptr` copy constructor is not atomic. There are `std::atomic_load` and `std::atomic_store` overloads for this case in c++11. 

But they are going to be deprecated in c++20 and the semantic will be changed to `std::atomic<std::shared_ptr<T>>`. http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2017/p0718r2.html#3.3

I bet we don't want to deal with different compilation issues tied to these changes.

We have no choice but to introduce kind of manual locking.